// @soulmachines/smwebsdk/docs/web-sdk/sidebar.js
/**
 * Returns an array of Category items for the API Reference section of the sidebar.
 * Each category is a type of member (class, interface, function, etc.) and
 * contains the members of that type.
 * 
 * Pages are generated from the api-extractor output markdown files.
 * 
 * @returns {import('@docusaurus/plugin-content-docs').SidebarItem[]}
 */
const apiReferenceItems = () => {

  const json = require('../../api-extractor-dist/smwebsdk.api.json');

  const categories = [];

  const libMembers = json.members[0].members;
  const libMembersByType = {};

  for (const member of libMembers) {
    if (!libMembersByType[member.kind]) {
      libMembersByType[member.kind] = [];
    }
    libMembersByType[member.kind].push(member);
  }

  Object.entries(libMembersByType).forEach(([kind, members]) => {
    const category = {
      type: 'category',
      label: kind,
      items: []
    };

    for (const member of members) {
      const memberItem = {
        type: 'doc',
        label: member.name,
        id: `api/smwebsdk.${member.name.toLowerCase()}`
      };

      category.items.push(memberItem);
    }

    categories.push(category);
  });

  return categories;
};

// @ts-check
/** @type {import('@docusaurus/plugin-content-docs').SidebarsConfig} */
module.exports = {
  websdkSidebar: [
    'overview',
    'getting-started',
    {
      type: 'category',
      label: 'Guides',
      link: {
        type: 'generated-index',
        title: 'WebSDK Guides',
        keywords: ['guides'],
        image: '/img/docusaurus.png',
      },
      collapsed: false,
      items: [
        {
          type: 'autogenerated',
          dirName: 'guides'
        },
      ],
    },
    {
      type: 'category',
      label: 'Troubleshooting',
      link: {
        type: 'generated-index',
        title: 'WebSDK Troubleshooting Guides',
        keywords: ['guides', 'troubleshooting'],
        image: '/img/docusaurus.png',
      },
      collapsed: true,
      items: [
        {
          type: 'autogenerated',
          dirName: 'troubleshooting'
        },
      ],
    },
    {
      type: 'category',
      label: 'API Reference',
      items: apiReferenceItems(),
    },
    {
      type: 'link',
      label: "Changelog",
      href: 'web-sdk/changelog'
    }
  ]
};
